source/mqtt_client_connection.c

```diff
@@ -99,6 +98,11 @@ napi_value aws_napi_mqtt_client_connection_close(napi_env env, napi_callback_inf
     aws_napi_callback_clean_up(&binding->on_connection_resumed);
     napi_delete_reference(env, binding->node_external);
 
+    /* destroy the native connection, which will destroy the subscriptions and
+     * queue destruction of all bound callbacks
+     */
+    aws_mqtt_client_connection_destroy(binding->connection);
```

Comment:
Bug 4: We were not destroying mqtt connections early enough in the cycle, so they could be holding callbacks in the sub trie, and that would prevent node from shutting down.